gr-uhd/include/gnuradio/uhd/rfnoc_replay.h

```diff
@@ -0,0 +1,157 @@
+     *               aligned to the size of the word in memory. Use get_word_size()
+     *               to get the memory word size.
+     * \param size Size of data to play back. This value must be aligned to the
+     *             size of the memory word and item size. Use get_word_size() to
+     *             get the memory word size and get_output_item_size() to get
+     *             the item size. This value will be used for the num_samps
+     *             component of the underlying stream command.
+     * \param port Which output port of the replay block to use
+     * \param time_spec Set the time for the first item. Any non-zero value is
+     *                  used to set the time in the header of the first packet.
+     *                  Most commonly, this is used to set the start time of a
+     *                  transmission. Note that this block will not wait for a
+     *                  time to occur, rather, it will tag the first outgoing
+     *                  packet with this time stamp.
+     * \param repeat Determines whether the data should be played repeatedly or
+     *               just once. If set to true, stop() must be called to stop
+     *               the play back.
+     * \throws uhd::value_error if offset+size exceeds the available memory.
+     * \throws uhd::op_failed Too many play commands are queued.
+     */
+    virtual void play(const uint64_t offset,
+                      const uint64_t size,
+                      const size_t port = 0,
+                      const ::uhd::time_spec_t time_spec = ::uhd::time_spec_t(0.0),
+                      const bool repeat = false) = 0;
+
+    /*! Stops playback
+     *
+     * Halts any currently executing play commands and cancels any other play commands
+     * that are waiting to be executed for that output port.
+     *
+     * \param port Which output port of the replay block to use
+     */
+    virtual void stop(const size_t port = 0) = 0;
+
+    /*! Sets the data type for items in the current record buffer for the given input
+     * port.
+     *
+     * \param type The data type
+     * \param port Which input port of the replay block to use
+     */
+    virtual void set_record_type(const std::string type, const size_t port = 0) = 0;
+
+    /*! Sets the data type for items in the current play buffer for the given output port.
+     *
+     * \param type The data type
+     * \param port Which output port of the replay block to use
+     */
+    virtual void set_play_type(const std::string type, const size_t port = 0) = 0;
+
+    /*! Sets the port for the next record
+     *
+     * \param port Which output port of the replay block to use
+     */
+    virtual void set_record_port(const size_t port) = 0;
+
+    /*! Sets the port for the next playback
+     *
+     * \param port Which output port of the replay block to use
+     */
+    virtual void set_play_port(const size_t port) = 0;
```

Comment:
I don't think you need the play port, record port, or play time APIs at all. They can all be passed in atomically with the command. This adds a statefulness that a) is not necessary and b) I really didn't understand it for the first minutes of reviewing this.

However, I would suggest adding an `issue_stream_command()` API as a substitute for the command handler API. You could potentially even use that from the command handler.

Reply:
Another way to view this: With this API, you have to first set the port (through the Python API) and then send a command (through the command API). That's a bit confusing.